### ISSUE TITLE
Nested colorize objects disable all attributes

### ISSUE THREAD

**Issue by Blacksmoke16** [crystal-lang/crystal#10652]
```cr
require "colorize"

foo = "foo".colorize.red

str = "hello #{foo} bar".colorize.underline

str => "\e[4mhello \e[31mfoo\e[0m bar\e[0m"
```

![image](https://user-images.githubusercontent.com/12136995/115802976-e2857980-a3ad-11eb-8412-cdf008e6d519.png)

Because coloring/modes are reset via `\e[0m` _all_ attributes are turned off, not just the one that is ending.  The correct code sequence should be like `"\e[4mhello \e[31mfoo\e[39m bar\e[24m"`.  I.e. `\e[39m` only resets the foreground color and `\e[24m` only resets the underline.

**Comment by HertzDevil** [comment]
That only solves part of the problem because you could still have:

```crystal
foo = "foo".colorize.red
str = "hello #{foo} bar".colorize.green
str
```

and `bar` will not be colored, even when `foo` ends with `\e[39`. The `#surround` method might handle this correctly, but not string interpolations I believe.

**Comment by straight-shoota** [comment]
Yeah and I don't think there can be a good solution really. To properly return to the previous colorize state, it would be necessary to keep track of the context.

This technically works, but the nested colorization needs to be actually nested inside the colorize `surround` context.
```cr
String.build do |io|
  Colorize.with.green.surround(io) do
    io << "hello #{"foo".colorize.red} bar"
  end
end # => "\e[32mhello \e[0;31mfoo\e[0;32m bar\e[0m"
```

Note that this is not concurrency safe because there is just a global `@@last_color` property.

**Comment by asterite** [comment]
I agree with @HertzDevil and @straight-shoota . The way these ansi codes work is not composable (unlike, say, html tags) because they are just toggles. And because of that, string interpolation won't work.

I think this is a "wontfix". Or, well, you can use `surround`.

**Comment by Blacksmoke16** [comment]
I mean I don't think we can solve for _all_ use cases so maybe this is more of an enhancement than a bug.  But is there any reason to _not_ use the correct toggles instead of `\e[0m`?  At least then it'll be more correct than before in more cases.
